nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java

```diff
@@ -228,6 +232,20 @@ protected void preProcessConfiguration(final Configuration config, final Process
         FsPermission.setUMask(config, new FsPermission(dfsUmask));
     }
 
+    @Override
+    protected void preProcessFileSystem(final FileSystem fileSystem, final ProcessContext context) throws IOException {
+        if (fileSystem instanceof DistributedFileSystem) {
+            final Path dirPath = new Path(context.getProperty(DIRECTORY).getValue());
```

Comment:
The `Directory` property indicates support for FlowFile Expression Language, which means that `getProperty()` needs to call `evaluateAttributeExpressions(FlowFile)`.  Since the FlowFile is not available in this method, it looks like a different approach is necessary.

One option is implementing this hook through a method called from `onTrigger()`, but that would log a warning for every invocation of the Processor.

Since the only current purpose of this hook is to log a warning, this may not be the best approach to the issue. Perhaps adding a note to the umask property would be a better option?

Reply:
> The `Directory` property indicates support for FlowFile Expression Language, which means that `getProperty()` needs to call `evaluateAttributeExpressions(FlowFile)`. Since the FlowFile is not available in this method, it looks like a different approach is necessary.
> 
> One option is implementing this hook through a method called from `onTrigger()`, but that would log a warning for every invocation of the Processor.
> 
> Since the only current purpose of this hook is to log a warning, this may not be the best approach to the issue. Perhaps adding a note to the umask property would be a better option?

Thanks. 

In addition to the doc update, I'd like to explore a different means to alert on this condition, if it is detected.  Having both should provide a helpful user experience.  I think I understand the constraints here; I'll give it another try.

Reply:
Changed the logic to throw on detection of this situation, as otherwise the file is written to HDFS with less restrictive permissions than are intended.